src/main/java/org/springframework/data/elasticsearch/repository/support/querybyexample/ExampleCriteriaMapper.java

```diff
@@ -0,0 +1,174 @@
+					criteria = applyPropertySpec(propertyPath + "." + key, value, exampleSpecAccessor, property, matchMode,
+							criteria);
+				}
+				continue;
+			}
+
+			criteria = applyPropertySpec(propertyPath, propertyValue, exampleSpecAccessor, property, matchMode, criteria);
+		}
+		return criteria;
+	}
+
+	private String getPropertyName(ElasticsearchPersistentProperty property) {
+		return property.isIdProperty() ? "_id" : property.getName();
+	}
+
+	private Criteria applyPropertySpec(String path, Object propertyValue, ExampleMatcherAccessor exampleSpecAccessor,
+			ElasticsearchPersistentProperty property, ExampleMatcher.MatchMode matchMode, Criteria criteria) {
+
+		if (exampleSpecAccessor.isIgnoreCaseForPath(path)) {
+			throw new InvalidDataAccessApiUsageException(
+					"Current implementation of Query-by-Example supports only case-sensitive matching.");
+		}
+
+		ExampleMatcher.StringMatcher stringMatcher = exampleSpecAccessor.getStringMatcherForPath(path);
+		if (!SUPPORTED_MATCHERS.contains(exampleSpecAccessor.getStringMatcherForPath(path))) {
+			throw new InvalidDataAccessApiUsageException(String.format(
+					"Current implementation of Query-by-Example does not support string matcher %s. Supported matchers are: %s.",
+					stringMatcher, SUPPORTED_MATCHERS));
+		}
+
+		final Object transformedValue = exampleSpecAccessor.getValueTransformerForPath(path)
+				.apply(Optional.ofNullable(propertyValue)).orElse(null);
+
+		if (transformedValue == null) {
+			criteria = tryToAppendMustNotSentence(criteria, path, exampleSpecAccessor);
+		} else {
+			if (property.isEntity()) {
+				return applyPropertySpecs(criteria, path, transformedValue,
+						mappingContext.getRequiredPersistentEntity(property), exampleSpecAccessor, matchMode);
+			} else {
+				return applyStringMatcher(applyMatchMode(criteria, path, matchMode), transformedValue, stringMatcher);
+			}
+		}
+		return criteria;
+	}
+
+	private Criteria tryToAppendMustNotSentence(Criteria criteria, String path,
+			ExampleMatcherAccessor exampleSpecAccessor) {
+		if (ExampleMatcher.NullHandler.INCLUDE.equals(exampleSpecAccessor.getNullHandler())
+				|| exampleSpecAccessor.hasPropertySpecifier(path)) {
+			return criteria.and(path).not().exists();
+		}
+		return criteria;
+	}
+
+	private Criteria applyMatchMode(Criteria criteria, String path, ExampleMatcher.MatchMode matchMode) {
+		if (matchMode == ExampleMatcher.MatchMode.ALL) {
+			return criteria.and(path);
+		} else {
+			return criteria.or(path);
+		}
+	}
+
+	private Criteria applyStringMatcher(Criteria criteria, Object value, ExampleMatcher.StringMatcher stringMatcher) {
+		return switch (stringMatcher) {
+			case DEFAULT, EXACT -> criteria.is(value);
+			case STARTING -> criteria.startsWith(validateString(value));
+			case ENDING -> criteria.endsWith(validateString(value));
+			case CONTAINING -> criteria.contains(validateString(value));
+			case REGEX -> throw new UnsupportedOperationException("REGEX matcher is unsupported");
+		};
```

Comment:
_I'm so glad we are on Java 17 finally (wish I could use Kotlin though)_

Reply:
I have added regex support on my upcoming changes